### PR TITLE
Use pluggy for plugin management

### PR DESCRIPTION
**Pull request by abkfenris**
[Pluggy](https://github.com/pytest-dev/pluggy) is the core of [py.test's ability to be extended](https://docs.pytest.org/en/latest/how-to/plugins.html), and it also is the plugin manager for [Tox](https://tox.wiki/en/latest/plugins.html), [Datasette](https://docs.datasette.io/en/latest/plugins.html), [Jupyter-FPS](https://github.com/jupyter-server/fps), [Napari](https://napari.org/stable/plugins/index.html), and [Conda](https://docs.conda.io/projects/conda/en/latest/dev-guide/plugins/index.html), among others.

In Xpublish a set of hooks is defined that plugins can implement, and a `pluggy.PluginManager` (as `xpublish.Rest.pm`) proxies requests to the plugins that have implemented the hooks and aggregates the results. 

Hooks define a set of possible kwargs that can be passed to downstream implementations, but not all implementations need to implement all of them (which makes it easy to add new kwargs without disrupting existing implementations). Hooks can also be defined as only returning the first response, or wrapping other hooks (dataset middleware?).

So far I've defined a handful of hooks in `xpublish.plugin.hooks:PluginSpec`:
- `app_router()`
- `dataset_router()`
- `get_datasets()`
- `get_dataset(dataset_id: str)`
- `register_hookspec()` - Which allows plugins to register new hook types

`get_datasets` and `get_dataset` allow plugins to provide datasets without loading them on launch, or overriding  `Rest._get_dataset_fn` or `Rest.setup_datasets()`.

I've kept the kwargs relatively minimal right now on the hooks as it's easier to expand the kwargs later, than it is to reduce them.

Other potential hooks include middleware, tabular or datatree datasets, a more end user focused UI.

I've additionally refactored the single dataset usage into it's own class `xpublish.SingleDatasetRest` to simplify some of the conditional logic which the accessor uses.

Example dataset provider plugin:

```py
class TestDatasetPlugin(xpublish.Plugin):
    name = "test-dataset-provider"

    @xpublish.hookimpl
    def get_datasets(self):
        return ["ww3"]

    @xpublish.hookimpl
    def get_dataset(self, dataset_id: str):
        if dataset_id == "ww3":
            return xr.open_dataset("../restful-grids/datasets/ww3_72_east_coast_2022041112.nc")
        
        return None
```

Pluggy references:
- https://pluggy.readthedocs.io/en/stable/
- https://docs.pytest.org/en/latest/how-to/writing_plugins.html
- https://docs.datasette.io/en/latest/writing_plugins.html
- https://docs.conda.io/projects/conda/en/latest/dev-guide/plugins/index.html

Replaces #145

### CODE REVIEW

**Comment by abkfenris** [comment]
PRs for the routers that I've made to work as pluggy based plugins:
- https://github.com/gulfofmaine/xpublish-opendap/pull/1
- https://github.com/gulfofmaine/xpublish-edr/pull/9

**Comment by abkfenris** [comment]
Other hooks that might be worth adding to Xpublish itself:
- FastAPI (things that need to hook into the FastAPI app or modify other routers) 
  - [Middleware](https://fastapi.tiangolo.com/tutorial/middleware/)
  - [Exception handlers](https://fastapi.tiangolo.com/tutorial/handling-errors/)
  - [Global dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/)
  - [Event handlers](https://fastapi.tiangolo.com/advanced/events/)
  - Dependency overrides?
  - [OpenAPI tags, and app info](https://fastapi.tiangolo.com/advanced/extending-openapi/)
  - Cache backend
- Dataset middleware? (I think pluggy already gives us it)

Hooks to delegate to plugins:
- UI
- Other datatypes (table/datatree)? Could also be part of Xpublish itself even if they aren't the focus.
- CLI
- [Config/settings loading/handling](https://fastapi.tiangolo.com/advanced/settings/)

**Comment by abkfenris** [comment]
After some pondering (powered by Future Island and a long commute in the dark), it seems like setting dependencies at the Plugin level is probably the wrong move. 

It makes it harder to override (say `get_dataset` if an app_router uses dataset_routers from other plugins, which is what I'm pondering on doing for forecasts) in some cases, without affecting how those dependencies are used in every case of a plugin. I believe dependencies set up this way also aren't usable from non routers.

When I get a chance, I'll explore passing in dependencies as arguments to `app_router` and `dataset_router` methods. I think that by passing them in as arguments, they may also be usable to the other plugin hook methods.

**Comment by abkfenris** [comment]
In a conversation with @mpiannucci I got to pondering how Dask fits into the plugin equation.

As far as I understand, xarray finds any dask client that's been initialized. If that's the case, it might be possible to add a new general `Plugin.setup()` hook, and an `xpublish-dask` plugin could create a client during setup. Then dataset loading plugins can implicitly use the Dask cluster.

If xarray can't take advantage of a dask client that's was initialized in another plugin due to namespace issues, then we could use a more explicit `Plugin.dask_client()` hook. That can be called and saved on the `Rest` instance, and provided to any `get_dataset()` hooks, for a signature of `get_dataset(id: str, client: dask.Client | None)`. Dataset loading plugins then could explicitly use the client as needed (and do things like smartly choosing `chunks`).

I think this could help take care of #54 in that the specifics of a Dask cluster are determined by the specifics of the loaded plugin. Hence we could have `xpublish-dask-local-cluster`, `xpublish-dask-coiled`, and the like to help configure different clusters.

File: tests/test_rest_api.py [hunk 1]
```diff
@@ -273,12 +280,18 @@ def test_ds_dict_cache(ds_dict):
...
+        # parameters is no longer generated when plugins use passed in deps
+        #  and get_dataset is replaced
+        assert len(openapi_schema['paths']['/']['get']['parameters']) == 0
+
+    with pytest.raises(KeyError):
+        # test cached value
+        # parameters is no longer generated when plugins use passed in deps
+        #  and get_dataset is replaced
+        openapi_schema = airtemp_rest.app.openapi()
+        assert len(openapi_schema['paths']['/']['get']['parameters']) == 0
```
Comment (abkfenris) [inline tests/test_rest_api.py:294]: With the change in dependency loading for built in routes (via plugins), `get_dataset` doesn't take an argument for single datasets, and thus the `parameters` key doesn't get added to the openapi spec. 

It still would get added (and need to be overridden) for any non-plugin routers, so the modified spec generator is still needed for single datasets.

File: xpublish/plugins/hooks.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        d.remove('__signature__')
+
+        return d
+
+
+class PluginSpec(Plugin):
+    """Plugin extension points"""
+
+    @hookspec
+    def app_router(self, deps: Dependencies) -> APIRouter:
```
Comment (jhamman) [inline xpublish/plugins/hooks.py:70]: Just for my learning, all these methods are optional right? I.e. a plugin could implement one or more of these hooks?

File: xpublish/plugins/included/dataset_info.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+from typing import List
```
Comment (jhamman) [inline xpublish/plugins/included/dataset_info.py:1]: question about file naming. Do we need the `/included/` part of the file path for the default plugins?

File: xpublish/plugins/manage.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+"""
+Load and configure Xpublish plugins from entry point group `xpublish.plugin`
+"""
+from importlib.metadata import entry_points
+from typing import Dict, Iterable, Optional
+
+from .hooks import Plugin
+
+
+def find_default_plugins(exclude_plugins: Optional[Iterable[str]] = None):
```
Comment (jhamman) [inline xpublish/plugins/manage.py:10]:
```suggestion
def find_default_plugins(exclude_plugins: Optional[Iterable[str]] = None) -> dict[str, Plugin]:
```

Is this right?

File: xpublish/plugins/manage.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    plugins: Dict[str, Plugin] = {}
+
+    for entry_point in entry_points()['xpublish.plugin']:
+        if entry_point.name not in exclude_plugins:
+            plugins[entry_point.name] = entry_point.load()
+
+    return plugins
+
+
+def load_default_plugins(exclude_plugins: Optional[Iterable[str]] = None):
```
Comment (jhamman) [inline xpublish/plugins/manage.py:26]: return type?

**Review by jhamman — APPROVED** [review]
This is awesome @abkfenris!!! Apologies for such a delayed review. 

A few questions below but they are mostly about just getting me caught up on what you (and @benbovy) have done here.

File: xpublish/plugins/hooks.py (hunk 2 above)
Comment (abkfenris) [reply xpublish/plugins/hooks.py:70]: Yep, they are all optional. I'm thinking I might explore mix-ins for common plugin patterns, in addition to a whole lot of documentation around plugins and how the whole ecosystem can work.

File: xpublish/plugins/included/dataset_info.py (hunk 3 above)
Comment (abkfenris) [reply xpublish/plugins/included/dataset_info.py:1]: I was wavering between `/default/` and `/included/` for the plugins that come with xpublish. I did want to keep them in the `/plugins/` directory, but I wanted to differentiate them from the plugin machinery. I initially had `/included_plugins/` or similar at the same level as `/plugins/` but that seemed like overkill and more confusing.

I'm also thinking that as part of documenting things, I'll probably make a few demo plugins that won't be automatically loaded via entrypoints, but can show off the different parts of the functionality (for example an dataset plugin that provides the xarray tutorial datasets), so I was thinking those would end up in `xpublish/plugins/demo/`

File: xpublish/plugins/manage.py (hunk 4 above)
Comment (abkfenris) [reply xpublish/plugins/manage.py:10]: Good catch. I had a nagging sensation that I actually needed a `Type[X]` in some project I was working on but I couldn't remember where!

Comment (abkfenris) [reply xpublish/plugins/manage.py:10]: Yep, it will be.

File: xpublish/plugins/manage.py (hunk 5 above)
Comment (abkfenris) [reply xpublish/plugins/manage.py]: I'm adding return types for all of `plugins/manage.py` and cleaning up the types of everything in plugins.

File: xpublish/plugins/included/dataset_info.py (hunk 3 above)
Comment (jhamman) [reply xpublish/plugins/included/dataset_info.py]: I think this is a bit overkill at this point but I'm fine with it.

**Comment by abkfenris** [comment]
Ok, @jhamman as soon as you are happy, this is ready for you to merge.

**Comment by jhamman** [comment]
Super excited by this @abkfenris!

**Comment by benbovy** [comment]
Well done @abkfenris, this is great! Sorry I didn't take the chance to look at it but happy to see it merged. I'll open new issues if I have any comment.

**Comment by abkfenris** [comment]
Thanks @jhamman and @benbovy 

@benbovy there is definitely more work to be done around plugins ([see comment above](https://github.com/xarray-contrib/xpublish/pull/146#issuecomment-1399541360), so there will be further iteration on them that I'd love feedback on. 

I'm hoping to work on the docs soon to at least give a rough intro to plugins, and explain the vision behind them/where Xpublish is going, so I'd love some feedback on that.

**Pull request merged**